File: submit_20176342_hw2/EasyTCPClient.go
/**
 * 20176342 Song Min Joon
 * EasyTCPClient.go
 **/

package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"
)

var serverName string = "nsl2.cau.ac.kr" //server host
var serverPort string = "26342"          //server port

func main() {
	c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM) // for exit the program gracefully
	go func() {
		for sig := range c {
			_ = sig
			byebye() // print byebye func
		}
	}()

	conn, err := net.Dial("tcp", serverName+":"+serverPort) //tcp connection

	if err != nil {
		//if server is not working, print and exit
		fmt.Printf("Please check your server is running\n")
		byebye()
		return
	}
	defer conn.Close()

	localAddr := conn.LocalAddr().(*net.TCPAddr) //get local port

	fmt.Printf("Client is running on port %d\n", localAddr.Port)

	for {
		//infinite input loop
		handleInput(conn)
	}

	defer conn.Close() // although when client gets panic, defer should disconnect socket gracefully
}

func byebye() {
	//print Bye bye~
	fmt.Printf("Bye bye~\n")
	os.Exit(0)
}

func handleInput(conn net.Conn) {
	/*
		this function prints Menu and 5 Options and wait for input Option.
	*/
	printOption()
	fmt.Printf("Please select your option :")
	var opt int
	fmt.Scanf("%d", &opt)
	processOption(opt, conn)
}

func processOption(opt int, conn net.Conn) {
	/*
		if option is given, it sends packet to server and get response.
	*/
	startTime := time.Now() //startTime for print RTT

	// var temp int
	// fmt.Scanf("%s", &temp)

	switch opt {
	case 1:
		// Option 1
		fmt.Printf("Input lowercase sentence: ")
		input, _ := bufio.NewReader(os.Stdin).ReadString('\n')
		startTime = time.Now()
		requestString := strconv.Itoa(opt) + "|" + input
		sendPacket(conn, requestString)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: %s\n", string(buffer[:bufferSize]))

	case 2:
		// Option 2
		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: client IP = %s, port = %s\n", string(strings.Split(string(buffer[:bufferSize]), ":")[0]), string(strings.Split(string(buffer[:bufferSize]), ":")[1]))

	case 3:
		// Option 3
		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		fmt.Printf("Reply from server: requests served = %s\n", string(buffer[:bufferSize]))

	case 4:
		// Option 4
		requestString := strconv.Itoa(opt) + "|"
		sendPacket(conn, requestString)
		buffer := make([]byte, 1024)
		bufferSize := readPacket(conn, &buffer)
		timeD, _ := time.ParseDuration(string(buffer[:bufferSize]))

		printDuration(timeD)

	case 5:
		// Option 5

		byebye()
		conn.Close()
		os.Exit(0)
	default:
		// not Option 1~5 (default)
		byebye()
		conn.Close()
		os.Exit(0)
	}
	printRTT(time.Since(startTime)) // print RTT Since startTime

}

func printRTT(d time.Duration) {
	//print RTT Time between before send packet and after send packet in Milliseconds.
	d = d.Round(time.Millisecond)

	s := d / time.Millisecond

	fmt.Printf("RTT = %dms \n\n\n", s)

}
func printDuration(d time.Duration) {
	//print server running time in proper form(HH:MM:ss)

	d = d.Round(time.Second)

	h := d / time.Hour
	d -= h * time.Hour

	m := d / time.Minute
	d -= m * time.Minute

	s := d / time.Second
	d -= s * time.Second

	fmt.Printf("Reply from server: run time = %02d:%02d:%02d\n", h, m, s)
}

func sendPacket(conn net.Conn, requestString string) {
	//send Packet to server
	conn.Write([]byte(requestString))
}

func readPacket(conn net.Conn, buffer *[]byte) int {
	//read Packet from server and saves to buffer and return buffer size.
	count, err := conn.Read(*buffer)
	if err != nil {
		fmt.Println("connection is closed by server")
		byebye()
		conn.Close()
		os.Exit(0)
	}
	return count
}

func printOption() {
	//print Menu and 5 Options.
	fmt.Printf("<Menu>\n")
	fmt.Printf("option 1) convert text to UPPER-case letters.\n")
	fmt.Printf("option 2) ask the server what the IP address and port number of the client is.\n")
	fmt.Printf("option 3) ask the server how many client requests(commands) it has served so far.\n")
	fmt.Printf("option 4) ask the server program how long it has been running for since it started.\n")
	fmt.Printf("option 5) exit client program\n")

}

File: submit_20176342_hw2/EasyTCPClient.java
/**
 * 20176342 Song Min Joon
 * EasyTCPClient.java
 **/

package submit_20176342_hw2;

import java.util.*;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.net.UnknownHostException;

public class EasyTCPClient {

    final static String serverName = "nsl2.cau.ac.kr"; // server host
    final static int serverPort = 26342;// server port

    public static class ByeByeThread extends Thread {
        // ByeBye Thread for graceful exit program.
        Socket conn;

        ByeByeThread(Socket conn) {
            this.conn = conn;
        }

        public void run() {
            try {
                Thread.sleep(200);
                byebye();

                try {
                    this.conn.close();
                } catch (IOException e) {

                }

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                e.printStackTrace();

            }
        }
    }

    public static void main(String[] args) {
        Socket conn = null;

        try {
            conn = new Socket(serverName, serverPort); // create TCP Socket
            Runtime.getRuntime().addShutdownHook(new ByeByeThread(conn)); // shutdown hook for graceful exit

            OutputStream os = conn.getOutputStream(); // output stream
            InputStream is = conn.getInputStream(); // input stream

            HandleInput H = new HandleInput(conn, os, is); // sub thread for input
            H.start(); //thread start

        } catch (UnknownHostException e) {
            //server not connected
            System.out.println("Please check your server is running");
            System.exit(0);
        } catch (IOException e) {
            //server not connected
            System.out.println("Please check your server is running");
            System.exit(0);

        }

    }

    public static void byebye() {
        //print bye bye
        System.out.println("Bye bye~");
    }

    public static void processOption(int opt, Socket conn, OutputStream os, InputStream is) {
        long startTime = new Date().getTime(); //startTime for print RTT

        try {
            String requestString;
            String reply;

            switch (opt) {
                case 1:
		            // Option 1

                    System.out.printf("Input lowercase sentence: ");
                    Scanner sc = new Scanner(System.in);
                    String input = sc.nextLine();
                    requestString = opt + "|" + input;

                    startTime = new Date().getTime();

                    sendPacket(os, requestString);

                    reply = readPacket(is);

                    System.out.printf("Reply from server: %s\n", reply);
                    break;
                case 2:
		            // Option 2

                    requestString = opt + "|";
                    sendPacket(os, requestString);
                    reply = readPacket(is);
                    System.out.printf("Reply from server: client IP = %s, port = %s\n", reply.split(":")[0],
                            reply.split(":")[1]);
                    break;
                case 3:
		            // Option 3

                    requestString = opt + "|";
                    sendPacket(os, requestString);
                    reply = readPacket(is);
                    System.out.printf("Reply from server: requests served = %s\n", reply);
                    break;
                case 4:
		            // Option 4

                    requestString = opt + "|";
                    sendPacket(os, requestString);
                    reply = readPacket(is);
                    printDuration(reply);
                    break;
                case 5:
		            // Option 5

                    conn.close();
                    System.exit(0);
                    break;
                default:
                    // not Option 1~5 (default)
                    conn.close();
                    System.exit(0);
                    break;

            }

        } catch (UnknownHostException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        printRTT(startTime);
    }

    public static void printRTT(long startTime) {
        //print RTT function

        System.out.printf("RTT = %dms\n\n\n", (new Date().getTime() - startTime));
    }

    public static void printDuration(String t) {
        //print server running time in proper form(HH:MM:ss)

        int h = 0;
        int m = 0;
        int s = 0;

        if (t.contains("h")) {
            h = Integer.parseInt(t.split("h")[0]);
            t = t.split("h")[1];
        }

        if (t.contains("m")) {
            m = Integer.parseInt(t.split("m")[0]);
            t = t.split("m")[1];
        }

        if (t.contains("s")) {
            s = Integer.parseInt(t.split("s")[0]);
        }

        System.out.printf("Reply from server: run time = %02d:%02d:%02d\n", h, m, s);
    }

    public static void sendPacket(OutputStream os, String requestString) {
        //use outputStream to send packet to server
        try {
            os.write(requestString.getBytes());
            os.flush();
        } catch (IOException e) {
            // do nothing
        }

    }

    public static String readPacket(InputStream is) {
	    //read Packet from server using InputStream.

        byte[] data = new byte[1024];

        try {
            int n = is.read(data);
            String res = new String(data, 0, n);
            return res;

        } catch (IOException e) {

            System.out.println("disconnected by server");
            System.exit(0);

            return "";
        }

    }

    public static void printOption() {
    	//print Menu and 5 Options.

        System.out.printf("<Menu>\n");
        System.out.printf("option 1) convert text to UPPER-case letters.\n");
        System.out.printf("option 2) ask the server what the IP address and port number of the client is.\n");
        System.out.printf("option 3) ask the server how many client requests(commands) it has served so far.\n");
        System.out.printf("option 4) ask the server program how long it has been running for since it started.\n");
        System.out.printf("option 5) exit client program\n");
    }

    static class HandleInput extends Thread {
        //HandleInput thread
        Socket conn;
        OutputStream os;
        InputStream is;

        HandleInput(Socket conn, OutputStream os, InputStream is) {
            this.conn = conn;
            this.os = os;
            this.is = is;
        }

        public void run() {
            while (true) {
                try {
                    //infinite loop
                    //print 5 options and take option input from user
                    printOption();
                    System.out.printf("Please select your option :");
                    Scanner sc = new Scanner(System.in);
                    int opt = sc.nextInt();
                    processOption(opt, this.conn, this.os, this.is);
                } catch (NoSuchElementException e) {
                    System.exit(0);
                }

            }

        }
    }
}